src/actions/mod.rs

```diff
@@ -306,20 +306,27 @@ impl InitActionContext {
         }
     }
 
-    fn init<O: Output>(&self, init_options: &InitializationOptions, out: &O) {
+    fn init<O: Output>(&self, init_options: InitializationOptions, out: &O) {
         let current_project = self.current_project.clone();
-        let config = self.config.clone();
-        // Spawn another thread since we're shelling out to Cargo and this can
-        // cause a non-trivial amount of time due to disk access
-        thread::spawn(move || {
-            let mut config = config.lock().unwrap();
-            if let Err(e) = config.infer_defaults(&current_project) {
-                debug!(
-                    "Encountered an error while trying to infer config defaults: {:?}",
-                    e
-                );
-            }
-        });
+
+        if let Some(config) = init_options.settings.map(|s| s.rust) {
+            self.config.lock().unwrap().update(config);
+        }
+
+        if self.config.lock().unwrap().needs_inference() {
```

Comment:
Coming from C++ I'd expect this line to `lock > read value > unlock` and then proceed to the inner block, however I'm not sure it always works that way - I've encountered a very similar case and surprisingly it resulted in a deadlock (I also tried to lock the same mutex in the inner scope again). Maybe because it desugared to a delayed drop of the temporary mutex guard?

To be sure and for clarity/reduced locking, could we extract it into something like...
```rust
let needs_inference = {
    let config = self.config.lock().unwrap();

    if let Some(cfg) = init_options.settings.map(|s| s.rust) {
        config.update(cfg);
    }
    config.needs_inference
};

if needs_inference {
...
```
?

Reply:
Sure